tensorzero-internal/src/model.rs

```diff
@@ -227,6 +229,131 @@ impl ModelConfig {
+                return Ok(cache_lookup);
+            }
+        }
+        let response = provider
+            .infer(
+                model_provider_request,
+                clients.http_client,
+                clients.credentials,
+            )
+            .instrument(span!(
+                Level::INFO,
+                "infer",
+                provider_name = model_provider_request.provider_name
+            ))
+            .await?;
+        // We already checked the cache above (and returned early if it was a hit), so this response was not from the cache
+        Ok(ModelInferenceResponse::new(
+            response,
+            model_provider_request.provider_name.into(),
+            false,
+        ))
+    }
+
+    /// Performs a streaming request to a specific provider, performing a cache lookup if enabled.
+    /// We apply model-provider timeouts to the future produced by this function
+    /// (as we want to apply the timeout to ClickHouse cache lookups).
+    ///
+    /// This function also includes a call to `peek_first_chunk` - this ensure that the
+    /// duration of the returned future includes the time taken to get the first chunk
+    /// from the model provider.
+    async fn streaming_provider_request<'request>(
+        &self,
+        model_provider_request: ModelProviderRequest<'request>,
+        provider: &'request ModelProvider,
+        clients: &'request InferenceClients<'request>,
+    ) -> Result<StreamResponseAndMessages, Error> {
+        // TODO: think about how to best handle errors here
+        if clients.cache_options.enabled.read() {
+            let cache_lookup = cache_lookup_streaming(
+                clients.clickhouse_connection_info,
+                model_provider_request,
+                clients.cache_options.max_age_s,
+            )
+            .await
+            .ok()
+            .flatten();
+            if let Some(cache_lookup) = cache_lookup {
+                return Ok(StreamResponseAndMessages {
+                    response: cache_lookup,
+                    messages: model_provider_request.request.messages.clone(),
+                });
+            }
+        }
+
+        let StreamAndRawRequest {
+            stream,
+            raw_request,
+        } = provider
+            .infer_stream(
+                model_provider_request,
+                clients.http_client,
+                clients.credentials,
+            )
+            .await?;
+
+        // Note - we cache the chunks here so that we store the raw model provider input and response chunks
+        // in the cache. We don't want this logic in `collect_chunks`, which would cause us to cache the result
+        // of higher-level transformations (e.g. dicl)
+        let mut stream = if clients.cache_options.enabled.write() {
+            let span = stream.span().clone();
+            // Note - it's fine to call `stream_with_cache_write` inside of `streaming_provider_request`,
```

Comment:
fine for who?

Reply:
I think it should be private and you should be forced to call the public API of the model layer